experimental/InjectiveResolutions/src/MonoidAlgebra.jl

```diff
@@ -0,0 +1,611 @@
+  return MonoidAlgebraElem(A, -underlying_element(b); check=false)
+end
+
+function (A::MonoidAlgebra{CoeffType})(c::CoeffType) where {CoeffType}
+  return MonoidAlgebraElem(A, A.algebra(c))
+end
+
+function (A::MonoidAlgebra)(c::Any)
+  return MonoidAlgebraElem(A, A.algebra(c))
+end
+
+function (A::MonoidAlgebra)(a::MonoidAlgebraElem)
+  @assert parent(a) === A
+  return a
+end
+
+function deepcopy_internal(a::MonoidAlgebraElem, dict::IdDict)
+  return parent(a)(deepcopy_internal(underlying_element(a), dict))
+end
+
+is_unit(a::MonoidAlgebraElem) = is_unit(underlying_element(a))
+
+function inv(a::MonoidAlgebraElem)
+  return MonoidAlgebraElem(parent(a), inv(underlying_element(a)))
+end
+
+monomial_basis(A::MonoidAlgebra, g::FinGenAbGroupElem) = monomial_basis(A.algebra, g)
+
+evaluate(a::MonoidAlgebraElem, vals::Vector) = evaluate(underlying_element(a),vals) 
+
+is_homogeneous(a::MonoidAlgebraElem) = is_homogeneous(underlying_element(a))
+
+function degree(
+        a::MonoidAlgebraElem{<:FieldElem, PT};
+        check::Bool=true
+    ) where {RT <: MPolyQuoRing, PT <: MonoidAlgebra{<:FieldElem, RT}}
+  !isdefined(a, :elem) && return zero(grading_group(parent(a)))
+  _a = underlying_element(a)
+  simplify(_a)
+  @req !iszero(_a) "Element must be non-zero"
+  return degree(_a.f; check)
+end
+
+function _degree_fast(a::MonoidAlgebraElem)
+  return _degree_fast(underlying_element(a))
+end
+
+function degree(::Type{Vector{Int}}, a::MonoidAlgebraElem; check::Bool=true)
+  _a = underlying_element(a)
+  @assert is_zm_graded((base_ring(parent(_a))))
+  d = degree(_a; check)
+  return Int[d[i] for i=1:ngens(parent(d))]
+end
+
+function dim(A::MonoidAlgebra)
+  return dim(A.algebra)
+end
+
+
+AbstractAlgebra.promote_rule(
+                             ::Type{CoeffType}, ::Type{T}
+                            ) where {CoeffType,T<:MonoidAlgebraElem{CoeffType}} = T
+
+# TODO: We would like to use the parametrization of the `MonoidAlgebra` 
+# with the type of its underlying ring directly. But this parameter only
+# provides the type of the ring, not its elements. So we have to work 
+# as if `MonoidAlgebra` was not parametrized.
+AbstractAlgebra.promote_rule(
+                             ::Type{MPolyDecRingElem{CoeffType, T}}, ::Type{U}
+                            ) where {CoeffType, T, U<:MonoidAlgebraElem{CoeffType}} = U
+
+AbstractAlgebra.promote_rule(
+                             ::Type{MPolyQuoRingElem{PolyType}}, ::Type{T}
+                            ) where {CoeffType, PolyType<:MPolyDecRingElem{CoeffType}, T<:MonoidAlgebraElem{CoeffType}} = T
+
+AbstractAlgebra.promote_rule(::Type{Int}, ::Type{T}) where {T<:MonoidAlgebraElem} = T
+AbstractAlgebra.promote_rule(::Type{ZZRingElem}, ::Type{T}) where {T<:MonoidAlgebraElem} = T
+
+gens(A::MonoidAlgebra) = [MonoidAlgebraElem(A, x) for x in gens(A.algebra)]
+number_of_generators(A::MonoidAlgebra) = ngens(A.algebra)
+getindex(A::MonoidAlgebra, i::Int) = MonoidAlgebraElem(A, A.algebra[i])
```

Comment:
This method is actually implemented generically; the actual function to implement is `gen`:
```suggestion
gen(A::MonoidAlgebra, i::Int) = MonoidAlgebraElem(A, gen(A.algebra, i))
```